### PR TITLE
Override configs on a single request level

### PR DESCRIPTION
**Pull request by wkal-pubnub**
Introduces `Endpoint.overrideConfiguration(PNConfiguration)` method for one-time, per-call overrides of certain PubNub  configuration options.

### CODE REVIEW

File: pubnub-gson/pubnub-gson-impl/src/integrationTest/java/com/pubnub/api/integration/PublishIntegrationTests.java [hunk 1]
```diff
@@ -107,6 +110,52 @@ public void testPublishMessageHistory() throws PubNubException {
...
+        try {
+            messagePayload.addProperty("name", "joe");
+            messagePayload.addProperty("age", 48);
+        } catch (JSONException e) {
+            e.printStackTrace();
+        }
+
+        final JsonObject whatToExpect = messagePayload;
+
+        PNConfiguration overrideConfig = PNConfigurationOverride.from(pubNub.getConfiguration()).setUserId(new UserId(expectedUser)).build();
```
Comment (marcin-cebo) [inline pubnub-gson/pubnub-gson-impl/src/integrationTest/java/com/pubnub/api/integration/PublishIntegrationTests.java:129]: I can do like this:
        PNConfiguration overrideConfig = PNConfigurationOverride.from(pubNub.getConfiguration())
                .setUserId(new UserId(expectedUser))
                .presenceTimeout(0)
                .heartbeatInterval(4)
                .cacheBusting(true)
                .build();
Do we want to allow modification of all methods from PNConfiguration#Builder?

File: pubnub-kotlin/pubnub-kotlin-api/src/main/kotlin/com/pubnub/api/Endpoint.kt [hunk 2]
```diff
@@ -1,5 +1,11 @@
 package com.pubnub.api
 
 import com.pubnub.api.endpoints.remoteaction.ExtendedRemoteAction
+import com.pubnub.api.v2.PNConfiguration
+import com.pubnub.api.v2.PNConfigurationOverride
 
-interface Endpoint<OUTPUT> : ExtendedRemoteAction<OUTPUT>
+interface Endpoint<OUTPUT> : ExtendedRemoteAction<OUTPUT> {
+    fun overrideConfiguration(action: PNConfigurationOverride.Builder.() -> Unit)
```
[the remark this answers is not in the document]
Comment (wkal-pubnub) [reply pubnub-kotlin/pubnub-kotlin-api/src/main/kotlin/com/pubnub/api/Endpoint.kt:8]: added the `from` version test, in addition to the `Builder.() -> Unit`

File: pubnub-kotlin/pubnub-kotlin-impl/src/integrationTest/kotlin/com/pubnub/api/integration/PublishIntegrationTests.kt [hunk 3]
```diff
@@ -161,6 +162,32 @@ class PublishIntegrationTests : BaseIntegrationTest() {
...
+        val messagePayload = generateMessage(pubnub)
+        val expectedUser = PubNub.generateUUID()
+
+        val observer = createPubNub {}
+        pubnub.test {
+            subscribe(expectedChannel)
+            observer.publish(
+                message = messagePayload,
+                channel = expectedChannel,
+            ).apply {
```
Comment (marcin-cebo) [inline pubnub-kotlin/pubnub-kotlin-impl/src/integrationTest/kotlin/com/pubnub/api/integration/PublishIntegrationTests.kt:177]: I was expecting to be able to do:
```
            observer.publish(
                message = messagePayload,
                channel = expectedChannel,
            ).overrideConfiguration(overwrittenConfiguration).sync()
```
Why not to add this possibility?

File: pubnub-gson/pubnub-gson-api/src/main/java/com/pubnub/api/endpoints/Endpoint.java [hunk 4]
```diff
@@ -4,5 +4,20 @@
...
 public interface Endpoint<T> extends ExtendedRemoteAction<T> {
+    /**
+     * Allows to override certain configuration options (see {@link com.pubnub.api.v2.BasePNConfigurationOverride.Builder}) for this request only.
+     * <p>
+     * {@link com.pubnub.api.v2.PNConfigurationOverride#from(com.pubnub.api.v2.BasePNConfiguration)} should be used to obtain a <code>PNConfigurationOverride.Builder</code>.
+     * Only options present in <code>PNConfigurationOverride.Builder</code> will be used for the override.
+     * <p>
+     * Example:
+     * <pre>
+     * configOverride = PNConfiguration.from(pubnub.configuration)
```
Comment (marcin-cebo) [inline pubnub-gson/pubnub-gson-api/src/main/java/com/pubnub/api/endpoints/Endpoint.java:15]: I can't find integtest that contains PNConfiguration.from
I would add one.
I found in test:
PNConfiguration overrideConfig = PNConfigurationOverride.from(pubNub.getConfiguration()).setUserId(new UserId(expectedUser)).build();

but this is not `PNConfiguration.from` as stated in KDoc above

File: pubnub-kotlin/pubnub-kotlin-impl/src/integrationTest/kotlin/com/pubnub/api/integration/PublishIntegrationTests.kt (hunk 3 above)
Comment (wkal-pubnub) [reply pubnub-kotlin/pubnub-kotlin-impl/src/integrationTest/kotlin/com/pubnub/api/integration/PublishIntegrationTests.kt:177]: added

File: pubnub-core/pubnub-core-impl/src/main/kotlin/com/pubnub/internal/EndpointCore.kt [hunk 5]
```diff
@@ -32,6 +33,11 @@ abstract class EndpointCore<Input, Output> protected constructor(protected val p
         final override val configuration: BasePNConfiguration
             get() = configOverride ?: pubnub.configuration
 
+        protected val retrofitManager: RetrofitManager
+            get() = configOverride?.let { override ->
```
[the remark this answers is not in the document]
Comment (wkal-pubnub) [reply pubnub-core/pubnub-core-impl/src/main/kotlin/com/pubnub/internal/EndpointCore.kt]: done

File: pubnub-kotlin/pubnub-kotlin-impl/src/main/kotlin/com/pubnub/internal/EndpointImpl.kt [hunk 6]
```diff
@@ -7,11 +7,13 @@ import com.pubnub.api.v2.PNConfigurationOverride
...
 
 abstract class EndpointImpl<OUTPUT>(private val endpoint: HasOverridableConfig) : Endpoint<OUTPUT> {
-    override fun overrideConfiguration(action: PNConfigurationOverride.Builder.() -> Unit) {
+    override fun overrideConfiguration(action: PNConfigurationOverride.Builder.() -> Unit): Endpoint<OUTPUT> {
         endpoint.overrideConfiguration(PNConfigurationImpl.Builder(endpoint.configuration).apply(action).build())
+        return this
     }
 
-    override fun overrideConfiguration(configuration: PNConfiguration) {
+    override fun overrideConfiguration(configuration: PNConfiguration): Endpoint<OUTPUT> {
```
Comment (marcin-cebo) [inline pubnub-kotlin/pubnub-kotlin-impl/src/main/kotlin/com/pubnub/internal/EndpointImpl.kt:15]: I would add/modify test to demonstrate usage of kind of builder pattern way of using overrideConfiguration

File: pubnub-gson/pubnub-gson-api/src/main/java/com/pubnub/api/endpoints/Endpoint.java (hunk 4 above)
Comment (wkal-pubnub) [reply pubnub-gson/pubnub-gson-api/src/main/java/com/pubnub/api/endpoints/Endpoint.java:15]: the kdoc is wrong, it should be PNConfigurationOverride, I'll change it

File: pubnub-kotlin/pubnub-kotlin-impl/src/main/kotlin/com/pubnub/internal/EndpointImpl.kt (hunk 6 above)
Comment (wkal-pubnub) [reply pubnub-kotlin/pubnub-kotlin-impl/src/main/kotlin/com/pubnub/internal/EndpointImpl.kt:15]: done

**Review by marcin-cebo — APPROVED** [review]

**Pull request merged**